### ISSUE TITLE
[API Proposal]: Expose xarch serialize instruction.

### ISSUE THREAD

**Issue by anthonycanino** [dotnet/runtime#66467]
### Background and motivation

Currently, the `cpuid` instruction also functions as a full serializing instruction --- one that ensures all modifications to flags, registers, and memory are drained before the next instruction can execute --- on x86 hardware. Executing a serializing instructions on P6 and more recent processor families constrain speculative execution because the results of speculatively executed instructions are discarded (see  `Chapter 8: Multiple-Processor Management` in the [Intel 64 and IA-32 Architectures Software Developer’s Manual, Volume 3A](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-3a-part-1-manual.html))

However, `cpuid` has overhead in that it clobbers `EAX`, `EBX`, `ECX`, and `EDX` registers. Intel exposes a new instruction, `serialize` under the `cpuid` feature flag `SERIALIZE` which acts as a full serializing instruction without the overhead of clobbering the registers as `cpuid` does. 

Full details of the `serialize` instruction can be found in `Chapter 2.1 Instruction Set Reference` in [Intel Architecture Instruction Set Extensions and Future Features](https://www.intel.com/content/dam/develop/external/us/en/documents/architecture-instruction-set-extensions-programming-reference.pdf).

### API Proposal

```C#
namespace System.Runtime.Intrinsics.X86
{
    /// <summary>
    /// This class provides access to Intel SERIALIZE hardware instruction via intrinsics
    /// </summary>
    [Intrinsic]
    [CLSCompliant(false)]
    public abstract class Serialize : X86Base
    {
        internal Serialize() { }

        public static new bool IsSupported { get => IsSupported; }

        [Intrinsic]
        public new abstract class X64 : X86Base.X64
        {
            internal X64() {  }

            public static new bool IsSupported { get => IsSupported; }
        }

        /// <summary>
        /// void _serialize (void);
        /// </summary>
        public static void Serialize() => Serialize();

    }
}
```

### API Usage

Broadly speaking, the developer can issue `Serialize.Serialize()` whenever they want to serialize the instruction stream before a branching conditional:

```C#
Serialize.Serialize();
if (someCondition)
{
  // ...
}
```

### Alternative Designs

dotnet developers currently need to call `System.Runtime.Intrinsics.X86.X86Base.CpuId()` to issue a full serializing instruction which has the extra overhead stated above as opposed to issuing the `serialize` instruction.

### Risks

As `serialize` functions as an alternative to `cpuid` with less overhead and a 0-arg use, we don't foresee any additional risks here.

**Comment by tannergooding** [comment]
Thanks for the proposal here! I've gone ahead and marked it ready for review.

Notably, we can't name the class and method both `Serialize` so we will need to disambiguate somewhere. Just an initial thought/suggestion is `X86Serialize.Serialize` which would also ensure the class isn't itself more generally conflicting.

I don't have a strong preference, however.

**Comment by anthonycanino** [comment]
Thanks for reviewing it! `X86Serialize.Serialize` is fine with me. I'll adjust.

**Comment by anthonycanino** [comment]
Hi @tannergooding , I can see this API hasn't been discussed yet (I believe) in the API proposal review meetings. 

Is there a long backlog of discussions before this would get discussed? Thanks!

**Comment by anthonycanino** [comment]
Thank you @tannergooding , I think this is ok for now. Thanks for moving it to the .NET 7 milestone.

**Comment by terrajobst** [comment]
* Looks good as proposed

```C#
namespace System.Runtime.Intrinsics.X86;

[Intrinsic]
[CLSCompliant(false)]
public abstract class X86Serialize : X86Base
{
    public static new bool IsSupported { get; }

    [Intrinsic]
    public new abstract class X64 : X86Base.X64
    {
        public static new bool IsSupported { get; }
    }

    public static void Serialize();
}
```

**Comment by tannergooding** [comment]
@anthonycanino, this is approved now and we can feel free to implement it whenever someone has availability.

**Comment by anthonycanino** [comment]
@tannergooding is adding the intrinsic lowering for the mono llvm path required for this implemtation, as seen in (https://github.com/dotnet/runtime/pull/61707) or is adding it to RyuJIT enough?

**Comment by tannergooding** [comment]
It shouldn't be required since this is a new ISA and `IsSupported` will report `false`.

However, it is generally preferable to bring both online where possible. CC. @fanyang-mono

**Comment by fanyang-mono** [comment]
Yes, if it is possible, it would be nice to add intrinsics support to Mono at the same time as CoreCLR. And I could answer questions if there is any, regarding to Mono.

**Comment by anthonycanino** [comment]
Would that include the mini and llvm lowering pathway of mono, or just one?

**Comment by anthonycanino** [comment]
Hi @tannergooding can you answer a question regarding adding a new ISA to RyuJIT? Currently, I am adding just two lines to the InstructionSetDesc.txt...

```
instructionset     ,X86   ,X86Serialize   ,   ,27 ,X86SERIALIZE ,x86serialize
; ...
implication        ,X86   ,X86SERIALIZE   ,X86Base
```

and generating with `gen.bat` in the directory. This change alone makes superpmi report a bunch of missing methods and method diffs, with no other adjustments. I assume it is because the JIT/EE interface has changed? Updating the jit guuid didn't fix the issue.

I'm trying to get the superpmi diffs stable so I can test a branch of mine that does implement the intrinsic.

**Comment by tannergooding** [comment]
Right, `superpmi` won't work as expected when you change the JIT/EE version.

I'm not sure what the best way to get a diff in this case is. CC. @dotnet/jit-contrib who may have advice here

**Comment by anthonycanino** [comment]
@AndyAyersMS will this work in my case though? Since I am creating a new ISA, I don't think I can isolate the base jit changes from the C# changes because of the re-generated corejit interface via InstructionSetDesc.txt. 

Essentially, I cannot only edit the jit due to the new ISA being needed for the JIT intrinsic nodes etc.

**Comment by AndyAyersMS** [comment]
Yes, it should work -- you would be comparing results from two entirely different builds.

**Comment by fanyang-mono** [comment]
> Would that include the mini and llvm lowering pathway of mono, or just one?

Both mini and LLVM.
